fetch-in-react-practice/client-ts/src/components/Todos.tsx

```diff
@@ -0,0 +1,151 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+import React, { useEffect, useState } from 'react';
+import PageTitle from './PageTitle';
+import TodoList from './TodoList';
+import TodoForm from './TodoForm';
+import { json } from 'stream/consumers';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export default function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function fetchTodos(): Promise<void> {
+      try {
+        const res = await fetch('./api/todos');
+        if(!res.ok) throw new Error(`fetch error ${res.status}`);
+        const todos = await res.json();
+        setTodos(todos);
+      } catch(err) {
+        setError(err)
+      } finally {
+        setIsLoading(false)
+      }
+    }
+    fetchTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo): Promise<void> {
+    try {
+      const res = await fetch('./api/todos', {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json"
+        },
+        body: JSON.stringify(newTodo)
+      });
+      const todo = await res.json();
+      setTodos((prev) => prev.concat(todo))
+    } catch(err) {
+      setError(err)
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todoId: number): Promise<void> {
+    let todoStatus = { isCompleted: false };
+    for (let i = 0; i < todos.length; i++) {
+      if (todoId === todos[i].todoId) {
+        todoStatus.isCompleted = !todos[i].isCompleted
+      }
+    }
```

Comment:
A lot of JS developers avoid for-loops like this when they can. You could rewrite this using forEach:
```
todos.forEach((todo) => {
  if (todoId === todo.todoId) {
    todoStatus.isCompleted = !todo.isCompleted;
  }
});
```

Another way to do it is to use `find`:
```
todoStatus.isCompleted = !todos.find((t) => t.todoId === todoId)?.isCompleted;
```

Reply:
These options are a lot less typing, I'll remember these too! I need to read up on `find` to fully understand how that one works.